### ISSUE TITLE
Ligature color-contrast checks result as incomplete

### ISSUE THREAD

**Issue by davyrwoo** [dequelabs/axe-core#1906]
Expectation: Ligature with failing contrast ratios should result as violations.

Actual: Ligature with failing contrast ratios result as incomplete.

Motivation: Ligature with inaccessible contrast should be considered violations, similar to how text color-contrast issues are violations.

![a11y ligature](https://user-images.githubusercontent.com/56053686/68982203-eecdd600-07ba-11ea-8d90-bb4454d3c772.PNG)

Example code:

```
<html>
	<style>
		.pass {
			color: #D7D7D7;
		}
		.fail {
			color: #AD2F42;
		}
		.dark {
			background: #201F1F
		}
	</style>
	<body class="dark">
		<main>
			<h1 class="pass">A11y test</h1>
			<div>
				<p class="pass"> &AElig </p>
				<p class="fail"> &AElig </p>
			</div>
		</main>
	</body>
</html>
```
axe-core version: 3.4.0
cypress-axe: 0.5.1
cypress: 3.6.1

**Comment by jeankaplansky** [comment]
No customer-facing docs required.

**Comment by padmavemulapati** [comment]
Verified this issue with the latest axe-coconut, I am still seeing the color-contrast issue in the "needs review" not in the voilations. 
@straker , can you please suggest me here, where can I see the fix.

**Comment by straker** [comment]
So the issue is our color-contrast check returns incomplete for the node because the content is too short: "Element content is too short to determine if it is actual text content." Not sure we can do anything about it as the code is working as intended.

**Comment by padmavemulapati** [comment]
As per @straker , nothing else here to do , so closing the issue

**Comment by KyleBastien** [comment]
@padmavemulapati and/or @straker would it be possible to check if the node only contains ligature's and then not care about the content length?

Seems like a big miss if all lone ligature's are unchecked for color contrast, since it's quite common to use ligature's for icon's, especially :heavy_check_mark: or :no_entry_sign: type icon's who might be green or red and should really be checked for color contrast.

**Comment by padmavemulapati** [comment]
@KyleBastien , I ran axe and axe-coconut on the given test file, when I observed with color-contrast analyzer, realised that red colored ligature is having low color-contrast ratio than the standard ratio.. so that issue is coming under needs review, which is expected as per the ticket. Please find below screenshots and please let me know if anything else need to be taken care inorder to evaluate this ticket.
![image](https://user-images.githubusercontent.com/47409579/70035502-f1ad3100-15d8-11ea-88fd-299ea67b6edc.png)
![ligature_in-red](https://user-images.githubusercontent.com/47409579/70035570-1b665800-15d9-11ea-82ce-89cec7b4ee11.PNG)

**Comment by KyleBastien** [comment]
@padmavemulapati I see that it's marked as needs review, but as per the original post on this issue, it should be a violation, not a incomplete, similar to all other text with contrast ratio issues. From the screenshot you shared, it still seems like it's marking as incomplete. If the length of the content is what is marking it as incomplete, then ligature's by the themselves shouldn't be bound to length, because they can contain just icon's by themselves and in those cases color contrast checked are very important.

**Comment by straker** [comment]
@KyleBastien So ✔️ or 🚫are not ligatures but emojis, which we cannot check for color-contrast at the moment (they have their own colors / look based on the platform, etc.). Text ligatures such as "AE" or "fi" should rarely be used on their own. The other case is a ligature icon font such as [Material Icon](https://material.io/resources/icons/) where the text "file" is replaced by an icon. These should be correctly check for color contrast.

**Comment by KyleBastien** [comment]
@straker I know ✔️ and 🚫are emoji's, I was trying to reference them as examples of types of icon ligatures that are frequently used by themselves.

> The other case is a ligature icon font such as Material Icon where the text "file" is replaced by an icon. These should be correctly checked for color contrast.

This doesn't appear to be the case, or there is a bug with this. See this example, which uses the [Office Fabric Ligature Icon's](https://developer.microsoft.com/en-us/fabric#/styles/web/icons):

https://2cl57.csb.app/

When I run axe-coconut on this I see this:

![image](https://user-images.githubusercontent.com/3861611/70080495-830daa80-15bb-11ea-86e0-8fc19e3e1c80.png)

But, when I run a color contrast tool on this, this doesn't pass:

![image](https://user-images.githubusercontent.com/3861611/70080705-e39ce780-15bb-11ea-8621-2164c2449fd3.png)

Let me know if I'm doing something wrong in this test, but this doesn't seem to check for color contrast on ligature icon's.

**Comment by straker** [comment]
So we filter unicode icons from being run through the color-contrast check if that's all the text includes. I believe we filter them since there is no guaranteed the unicode icon will display properly, However, in this case the unicode character is being used as a key for the fonts ligature symbols, but I'm not sure we can detect that.

**Comment by KyleBastien** [comment]
@straker Can you clarify more on this?

> I believe we filter them since there is no guaranteed the unicode icon will display properly

When you say display properly here, what do you mean?

**Comment by straker** [comment]
The end result is not the unicode character but a bunch of boxes as the OS doesn't support that range 
![image](https://user-images.githubusercontent.com/2433219/70095975-3fc73200-15e3-11ea-8305-47a18e6e53c6.png)

**Comment by KyleBastien** [comment]
@straker Ahh thanks for clarifying.

Would it be possible to turn off skipping over Unicode characters via a config to axe-core?

We won't be testing ligature's on OS's that don't render them, but I also understand that might be specific to our use case. So I'm wondering if a config might be possible here?

**Comment by straker** [comment]
@KyleBastien Sure. Would you be interested in writing a PR for it?

**Comment by KyleBastien** [comment]
@straker Sure! Could you point me towards where the logic is that currently skips over Unicode characters?

And also where to add a new config property?

**Comment by straker** [comment]
Awesome!

So there will be 2 things that need to happen. 1st, we need to stop filtering unicode and 2nd, we need to have an option to ignore the length requirement (as a single unicode will still be "too short").

The first part is done in the [color-contrast-matcher function](https://github.com/dequelabs/axe-core/blob/develop/lib/rules/color-contrast-matches.js#L97-L106) (the `nonBmp ` flag). Matchers can't have options, so we'll just going to allow unicode through as they can be used in different ways.

For the second part, options can be added to individual checks (what a rule runes). You can see an option for [aria-roledescription](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/aria-roledescription.json#L4-L14) as an example.  So you'll need to add an `option` object to the [color-contrast check](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/color/color-contrast.json). We'll need two options: 1 for ignoring unicode (on by default) and 1 for ignoring length requirement (off by default)

Lastly, you'll need to use the option in the [check function itself](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/color/color-contrast.js). The check function is passed an `options` object that you can use. The check should return true if the text only is unicode (if the option is on) and [then not ignore length](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/color/color-contrast.js#L35-L38).

**Comment by KyleBastien** [comment]
@straker I guess I was thinking of something slightly different for the length requirement piece that might not need an additional flag.

Basically, as I understand from above, the length requirement doesn't come into play for ligatures when those ligatures are by themselves, since Material Icon ligatures appear to be checked correctly for color contrast. Am I correct there?

Could we treat unicode characters the same, if the "filterUnicodeCharacters" is set to false? i.e. If it is just a unicode character by itself, it's probably an icon therefore do a color contrast check independent on length?

Also, if I'm wrong in my understanding of how ligatures are checked, what do you think of making the flag something like "ignoreLigatureLength", so this is more scoped to just ligature checks and not ignore length checks for everything?

**Comment by straker** [comment]
Material Icon uses text as a ligature key so the string "android" is replaced by the Android icon. That's why it works for our color contrast because it uses normal text in the DOM. @WilcoFiers what do you think about @KyleBastien suggestions?

**Comment by KyleBastien** [comment]
@straker 

I took a first shot at the implementation you described, using two config values.

https://github.com/KyleBastien/axe-core/commit/58e9dea49d29203f4e78e5591595068da929ede0

If you could take a look and let me know if you think this is along the right lines, that would be greatly appreciated!

**Comment by KyleBastien** [comment]
@straker Happy New Year!

If you get the opportunity to take a look @ the change I implemented above, I would greatly appreciate it!

Thank you!

**Comment by straker** [comment]
@KyleBastien Happy New Year to you as well!

Looking over the changes it looks like everything's working. I don't have any immediate concerns from my quick look, so feel free to open a PR.

**Comment by KyleBastien** [comment]
@straker Thanks! I opened the PR!
